aten/src/ATen/native/cuda/Distributions.cu

```diff
@@ -475,6 +475,30 @@ void random_kernel_cuda(TensorIterator& iter, uint64_t range, int64_t base, Gene
    });
 }
 
+void normal_kernel_cuda(TensorIterator& iter, double mean_, double std_, Generator* gen_) {
```

Comment:
OK, so this is the second time I'm reviewing the transform from `GENERATE_KERNEL2` to this code, and the thing that I found a bit remarkable here, is that the code has gotten a lot longer. Is this fundamental to how `distribution_nullary_kernel` works or is there something we can factor out here to reduce how much code we have to write in all of these cases?

Reply:
I think following are what looks analogous to me with the old code:
```
distribution_elementwise_grid_stride_kernel -> GENERATE_KERNEL2
overloads of normal_cuda/normal_out_cuda -> dispatches in CUDAType::_th_normal_out in CUDAType.cpp
*kernel_cuda -> THCTensor_(*) code in THCTensorRandom.cu
```
The added bit of code is `distribution_nullary_kernel` which handles the TensorIterator stuff. Original plan was to modify Loops.cuh to have an `elementwise_kernel` and `launch_kernel` specialized to run in grid-stride mode, but the added complexity of distributions needing a transform lambda, the idx value to do unrolling and recursively calculating the philox offset (at the beginning of distribution_nullary_kernel) made me copy bits from `gpu_nullary_kernel` code.

As the for the std::is_same-s, even though code generation with the macros look less line of codes, I am usually against the macros as it is very unreadable. Like, it took a lot of bookkeeping to get every types right just by looking at the code. I do agree with you on the template specialization on the is_same's though. For now it just seemed easy to do.
```
GENERATE_KERNEL2(generate_normal, float, double mean, double stdv, float, curand_normal, (x * stdv) + mean)
GENERATE_KERNEL2(generate_normal, double, double mean, double stdv, double, curand_normal_double, (x * stdv) + mean)

GENERATE_KERNEL1(generate_exponential, float, double lambda, float, curand_uniform, (float)(-1. / lambda * log(x)))
GENERATE_KERNEL1(generate_exponential, double, double lambda, double, curand_uniform_double, (double)(-1. / lambda * log(x)))

GENERATE_KERNEL2(generate_cauchy, float, double median, double sigma, float, curand_uniform, (float)(median + sigma * tan(M_PI*(x-0.5))))
GENERATE_KERNEL2(generate_cauchy, double, double median, double sigma, double, curand_uniform_double, (double)(median + sigma * tan(M_PI*(x-0.5))))

GENERATE_KERNEL2(generate_normal, at::Half, double mean, double stdv, float, curand_normal, (ScalarConvert<float, at::Half>::to((x * stdv) + mean)))
GENERATE_KERNEL1(generate_exponential, at::Half, double lambda, float, curand_uniform, (ScalarConvert<float, at::Half>::to((float)(-1. / lambda * log(x)))))
GENERATE_KERNEL2(generate_cauchy, at::Half, double median, double sigma, float, curand_uniform, (ScalarConvert<float, at::Half>::to((float)(median + sigma * tan(M_PI*(x-0.5))))))

#include <THC/generic/THCTensorRandom.cu>
#include <THC/THCGenerateAllTypes.h>

#include <THC/generic/THCTensorRandom.cu>
#include <THC/THCGenerateBoolType.h>

#undef GENERATE_KERNEL1
#undef GENERATE_KERNEL2
.
.
.
#if defined(THC_REAL_IS_DOUBLE)
GENERATE_KERNEL1(generate_geometric, double, double p, double, curand_uniform_double, ceil(log(x) / log(1-p)))
#else
GENERATE_KERNEL1(generate_geometric, scalar_t, double p, float, curand_uniform, (ScalarConvert<float, scalar_t>::to(ceilf(logf(x) / log(1-p)))))
#endif
#if defined(THC_REAL_IS_LONG) || defined(THC_REAL_IS_DOUBLE) || defined(THC_REAL_IS_FLOAT)
#define CURAND64(STATE) (((uint64_t)curand(STATE)) << 32) | (uint64_t)curand(STATE)
GENERATE_KERNEL2(generate_random, scalar_t, int32_t base, uint32_t range, uint32_t, curand, \
    static_cast<scalar_t>(static_cast<int32_t>((x % range) + base)))
GENERATE_KERNEL2(generate_random_64, scalar_t, int64_t base, uint64_t range, uint64_t, CURAND64, \
    static_cast<scalar_t>(static_cast<int64_t>((x % range) + base)))
#elif defined(THC_REAL_IS_HALF)
GENERATE_KERNEL2(generate_random, scalar_t, int32_t base, uint32_t range, uint32_t, curand,
    (ScalarConvert<int32_t, scalar_t>::to(static_cast<int32_t>(x % range + base))))
#else
GENERATE_KERNEL2(generate_random, scalar_t, int32_t base, uint32_t range, uint32_t, curand,
    static_cast<scalar_t>(static_cast<int32_t>(x % range + base)))
#endif
```

Reply:
👍 SGTM